### DISCUSSION TITLE
"next dev" and "next build" taking disproportionately long to execute

### DISCUSSION THREAD

**Discussion by DopamineDriven** [vercel/next.js#20362] [Help]
Nextjs projects are taking an exceedingly long amount of time to build both locally and on vercel. This all started with a `Node.js heap out of memory` error on Friday morning. It gets to the next build step after installing dependencies then it stalls for ~10-15 minutes before finally proceeding with the build. Builds have usually only taken 2-3 minutes tops to complete. Here is the[ url for the repo in question](https://github.com/DopamineDriven/drisdell-consulting-services). I've included my OS and device specs below:
![about-your-pc](https://user-images.githubusercontent.com/46355797/102783311-56603880-4360-11eb-8fc6-973cc5fde6c4.png)

Headless WP via Amazon LightSail -- Linux instance ft Bitnami and Automattic (cloudfront CDN) -- using WPGraphQL to communicate with Headless CMS endpoint

Package.json for this project
```js
{
	"name": "drisdell-consulting-services",
	"version": "1.0.0",
	"main": "index.js",
	"repository": "https://github.com/DopamineDriven/drisdell-consulting-services.git",
	"author": "Andrew Ross <andrew.simpson.ross@gmail.com>",
	"license": "MIT",
	"private": true,
	"scripts": {
		"dev": "next dev && NODE_OPTIONS=--inspect",
		"build": "next build NODE_OPTIONS=max-old-space-size=8192",
		"start": "next start",
		"analyze": "cross-env ANALYZE=true yarn build",
		"tsconfig:effective": "tsc --showConfig",
		"postinstall": "npx patch-package",
		"client:codegen": "yarn apollo codegen:generate --localSchemaFile=graphql.schema.json --target=typescript --includes=./lib/graphql/**/*.ts --tagName=gql --addTypename --globalTypesFile=_types/graphql-global-types.ts",
		"codegen:generate": "graphql-codegen -r dotenv/config --config codegen.yml",
		"codegen:combo": "yarn codegen:generate && yarn client:codegen"
	},
	"husky": {
		"hooks": {
			"pre-commit": "yarn tsc && pretty-quick --staged && yarn tsconfig:effective"
		}
	},
	"dependencies": {
		"@apollo/client": "^3.3.6",
		"@fortawesome/fontawesome-svg-core": "^1.2.32",
		"@fortawesome/free-brands-svg-icons": "^5.15.1",
		"@fortawesome/react-fontawesome": "^0.1.12",
		"@headlessui/react": "^0.2.0",
		"@react-aria/overlays": "^3.6.0",
		"@tailwindcss/aspect-ratio": "^0.2.0",
		"@tailwindcss/forms": "^0.2.1",
		"@tailwindcss/typography": "^0.3.1",
		"@tailwindcss/ui": "^0.7.2",
		"@tailwindui/react": "^0.1.1",
		"apollo": "^2.32.0",
		"classnames": "^2.2.6",
		"cross-env": "^7.0.3",
		"date-fns": "^2.16.1",
		"graphql": "^15.4.0",
		"js-cookie": "^2.2.1",
		"keen-slider": "^5.3.5",
		"lodash.random": "^3.2.0",
		"lodash.throttle": "^4.1.1",
		"next": "^10.0.4-canary.8",
		"next-mdx-remote": "^1.0.1",
		"react": "^17.0.1",
		"react-aria": "^3.1.1",
		"react-dom": "^17.0.1",
		"react-ga": "^3.2.1",
		"react-intersection-observer": "^8.30.3",
		"react-markdown": "^5.0.2",
		"react-merge-refs": "^1.1.0",
		"react-syntax-highlighter": "^15.3.0",
		"tailwindcss": "^2.0.2",
		"tailwindcss-gradients": "^3.0.0",
		"vercel": "^21.0.0"
	},
	"devDependencies": {
		"@graphql-codegen/cli": "^1.19.4",
		"@graphql-codegen/introspection": "^1.18.0",
		"@graphql-codegen/typed-document-node": "^1.17.10",
		"@graphql-codegen/typescript": "^1.17.11",
		"@graphql-codegen/typescript-apollo-client-helpers": "^1.1.0",
		"@graphql-codegen/typescript-operations": "^1.17.10",
		"@graphql-codegen/typescript-react-apollo": "^2.1.1",
		"@next/bundle-analyzer": "^10.0.3",
		"@types/classnames": "^2.2.11",
		"@types/gtag.js": "^0.0.3",
		"@types/js-cookie": "^2.2.6",
		"@types/lodash.random": "^3.2.6",
		"@types/lodash.throttle": "^4.1.6",
		"@types/node": "^14.14.8",
		"@types/react": "^16.9.56",
		"@types/react-dom": "^16.9.9",
		"@types/react-syntax-highlighter": "^13.5.0",
		"@vercel/node": "^1.8.5",
		"autoprefixer": "^10.1.0",
		"graphql-tag": "^2.11.0",
		"husky": "^4.3.0",
		"jest": "^26.6.3",
		"next-compose-plugins": "^2.2.1",
		"npm-run-all": "^4.1.5",
		"patch-package": "^6.2.2",
		"postcss": "^8.2.1",
		"postcss-flexbugs-fixes": "^5.0.1",
		"postcss-preset-env": "^6.7.0",
		"postinstall-postinstall": "^2.1.0",
		"prettier": "^2.2.1",
		"pretty-quick": "^3.1.0",
		"react-test-renderer": "^17.0.1",
		"stylelint": "^13.8.0",
		"stylelint-config-recommended": "^3.0.0",
		"typescript": "^4.1.3",
		"yaml-loader": "^0.6.0"
	},
	"browserslist": [
		">0.3%",
		"not ie 11",
		"not dead",
		"not op_mini all"
	]
}

```

Example Vercel Build output since this issue has started: 
![build](https://user-images.githubusercontent.com/46355797/102784342-e05cd100-4361-11eb-89ed-d65745edf5ba.png)

I have been trouble shooting this issue since Friday morning and opened an [issue on Stack Overflow](https://stackoverflow.com/questions/65375962/nextjs-yarn-amplify-causing-memory-overrages) on Saturday evening. I have tried everything from defragmenting, to setting a new NODE_OPTIONS system environmental variable, to rolling back to previous versions of Node (v15->v14) and Next@10.0.x (as well as Next@10.0.4-canary-8)

I have also tried running `yarn cache clean --mirror` to remove global cache files. If I run `yarn cache clean --all` it will remove all cached files, both locally and globally...I'm not entirely certain as to what to do next. 

I rolled back an update for windows I installed on 12/18/20 as well. My next option is to salvage my files and restore factory settings...

**Comment by vladpereverzyev** [comment]
There are a few things in this configuration that I would investigate before considering a Windows reinstall.

The `Node.js heap out of memory` error is important, but the 10–15 minute delay suggests that the process may be spending a significant amount of time under memory pressure rather than simply failing because the heap limit is too low.

One thing that immediately stands out is the build script:

```json
"build": "next build NODE_OPTIONS=max-old-space-size=8192"
```

`NODE_OPTIONS` is a Node environment variable and should not be passed to `next build` like this. If the intention is to increase the Node heap, use something like:

```json
"build": "cross-env NODE_OPTIONS=--max-old-space-size=8192 next build"
```

I would also avoid the canary version while troubleshooting:

```json
"next": "^10.0.4-canary.8"
```

Use a stable Next.js release and make sure the exact version is locked in `yarn.lock`.

There is another important point: because the project uses WPGraphQL/Apollo and apparently performs data fetching during the Next.js build, I would check whether `getStaticProps`, `getStaticPaths`, or any other build-time GraphQL requests are hanging or becoming extremely slow.

For example, temporarily add timing around the GraphQL requests:

```js
console.time('graphql')
const result = await client.query(...)
console.timeEnd('graphql')
```

This would tell us whether the 10–15 minutes are actually being spent waiting for the WordPress/GraphQL endpoint.

I would also run a completely clean build:

```bash
rm -rf node_modules .next
yarn cache clean
yarn install --frozen-lockfile
yarn build
```

On Windows, remove `node_modules` and `.next` manually or with PowerShell.

Then compare:

```bash
node -v
yarn -v
yarn why next
```

I would also check whether the same commit produces the same behavior in a clean environment, for example Vercel or a fresh Node installation. Since the problem occurs both locally and on Vercel, that makes a machine-specific Windows issue less likely.

One other detail I'd investigate is the `dev` script:

```json
"dev": "next dev && NODE_OPTIONS=--inspect"
```

This does not enable the Node inspector for `next dev` as intended. If debugging is required, `NODE_OPTIONS` needs to be set before starting Next.js.

So I would not factory-reset the machine yet. I'd first isolate whether the slowdown is caused by:

1. Node heap/GC pressure
2. build-time GraphQL requests
3. the canary Next.js version
4. a dependency/build plugin
5. or the local environment itself

If you can share the actual `next build` output around the point where it stalls, that would be the most useful next piece of information.

**Comment by js713-lab** [comment]
### When `next dev` / `next build` is extremely slow

Common fixes that actually move the needle:

1. **Upgrade Node** to the version Next recommends for your release; old Node + large graphs is painful.
2. Clear caches: `rm -rf .next node_modules/.cache` then reinstall.
3. Check for accidental huge imports (barrel files pulling entire icon/UI libs into every route).
4. On Vercel, watch for OOM — increase memory or reduce concurrency; locally set `NODE_OPTIONS=--max-old-space-size=8192` only as a bridge.
5. Disable unused experimental flags; try without Turbopack once to compare.
6. `next build --debug` / analyze bundle to find oversized modules.

If you share Next version + whether it’s local or Vercel + approximate `node_modules` size, we can narrow it.
